rxjs-1/carsLeaderBoard.js

```diff
@@ -0,0 +1,135 @@
+const { combineLatest } = require('rxjs');
+const { map, bufferCount } = require("rxjs/operators");
+const { getCarSpeed } = require('./carSpeed');
+const { eventToObservable } = require('./eventToObservable')
+
+// return the table sorted
+// using the key paramters
+// note : the key params as to be a string : we can improve this part
+const sortTableObject = (carObjects, key) => {
+  return carObjects.sort((a, b) => {
+      const x = a[key];
+      const y = b[key];
+      return ((x < y) ? -1 : ((x > y) ? 1 : 0));
+  });
+}
+
+// check if the car exist using it's name
+const carExistInBoard = (board, carName) => {
+  const indexOfCar = (boardObj) => boardObj.carName === carName;
+  return (board.findIndex(indexOfCar) >= 0);
+}
+
+// return a car object
+const baseCarObjects = (cars) => {
+  // return the array of premade leaderboard data
+  return cars.map((car) => {
+    return {
+      time: car.time,
+      carName: car.carName,
+      xLocation: car.xLocation,
+      speed: 0
+    }
+  });
+}
+
+// we prepare the board raw data in order to complete the computation
+const prepareDataProcessBoard = (board) => {
+  let pos = board.length;
+  // we order by inverted position
+  // in order to simplify the calcul of the gap
+  return sortTableObject(board, 'xLocation').map((car) => {
+      const ret = {
+        ...car,
+        position: pos,
+        leaderGapDistance: 0,
+        leaderGapTime: 0,
+        speed: car.speed
+
+      };
+      pos--;
+      return ret;
+    });
+  return
+}
+
+// finalize the data of the board comptuation
+const processBoard = (board) => {
+  const retBoard = prepareDataProcessBoard(board);
+  // we process with the calculation of the inverted list
+  retBoard.forEach((car, index) => {
+    if (index < retBoard.length) {
+      const nextCar = retBoard[index+1];
+      // just to be sure ...
+      if (nextCar) {
+        car.leaderGapDistance = nextCar.xLocation - car.xLocation;
```

Comment:
An immutable helper to do the same thing (i.e. return a new car with updated `leaderGapDistance ` key) would even be better IMO. With that you could use a `map` instead of a `forEach` and return the result directly:
```typescript
// something like
return retBoard.map((car, index) => {
  ...
  return setLeaderGapDistance(car, newCar);
  ...
});
```